core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2673,27 +2673,36 @@ class KafkaApis(val requestChannel: RequestChannel,
     val groupId = offsetDeleteRequest.data.groupId
 
     if (authorize(request, DELETE, GROUP, groupId)) {
+      val authorizedTopics = filterAuthorized(request, READ, TOPIC,
+        offsetDeleteRequest.data.topics.asScala.map(_.name).toSeq)
+
       val topicPartitions = offsetDeleteRequest.data.topics.asScala.flatMap { topic =>
         topic.partitions.asScala.map { partition =>
-          new TopicPartition(topic.name, partition.partitionIndex)
+          val tp = new TopicPartition(topic.name, partition.partitionIndex)
+          val error = if (!authorizedTopics.contains(topic.name))
+            Errors.TOPIC_AUTHORIZATION_FAILED
+          else if (!metadataCache.contains(tp))
+            Errors.UNKNOWN_TOPIC_OR_PARTITION
+          else
+            Errors.NONE
+          tp -> error
         }
-      }.toSeq
-
-      val authorizedTopics = filterAuthorized(request, READ, TOPIC, topicPartitions.map(_.topic))
-      val (authorizedTopicPartitions, unauthorizedTopicPartitions) = topicPartitions.partition { topicPartition =>
-        authorizedTopics.contains(topicPartition.topic)
       }
 
-      val unauthorizedTopicPartitionsErrors = unauthorizedTopicPartitions.map(_ -> Errors.TOPIC_AUTHORIZATION_FAILED)
-      val (groupError, authorizedTopicPartitionsErrors) = groupCoordinator.handleDeleteOffsets(groupId, authorizedTopicPartitions)
-      val topicPartitionsErrors = unauthorizedTopicPartitionsErrors ++ authorizedTopicPartitionsErrors
+      val (authorizedAndKnowTopicPartitions, unauthorizedOrUnknownTopicPartitionErrors) =
```

Comment:
This is an established pattern, but these intermediate collections are a little irritating. Alternatively we could build the collections we need in the pass above where we check authorization and existence. Not that big of a deal though since it's probably a rarely used api.

Reply:
I have changed this.